### PR TITLE
Add async writes

### PR DESCRIPTION
**Pull request by tomekl007**

### CODE REVIEW

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java
Comment (dimas-b) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: Is this not what logged batches provide?

Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: yes, this is why we allow logged batch (`atomic` directive) or the new `async`.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (dimas-b) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java:185]: Well, since `@async` annotates specific mutations, we should be able to use a different return type. Put it another way, making a mutation async is a schema change from the client's perspective, so it should not be a surprise the user that the return value is different. WDYT?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java
Comment (dimas-b) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: Yes, but why not allow both? If the client does not want to wait for the acknowledgement of a batch write, it's the client's choice. Why would we want to explicitly forbid it?

Comment (dougwettlaufer) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: +1 an async batch makes sense to me.

File: testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql
Comment (dougwettlaufer) [inline testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql:1]: These tests should probably make use of @olim7t change in https://github.com/stargate/stargate/pull/906

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java:185]: As far as I can see, we don't have the knowledge of the semantic of the GraphQL query at the `GraphqlResourceBase` where we decide about the HTTP response codes. Also, all responses all returned with the standard `200` code.
We may be able to refactor the whole `GraphQL` worklow to give the `GraphqlResourceBase`:
```
   .thenApply(result -> asyncResponse.resume(result.toSpecification()))
```
feedback about whether it should return `202` or `200` but I am not sure if this is worth the effort (@olim7t wdyt?)
I think that a good middle ground is returning `accepted` in the GraphQL response body as @dimas-b suggested.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: Here is my reasoning around forbidding it:

On the one hand, we may allow both operations - it seems that batch without waiting for the `ack`. On the other hand, I think that we should enforce good patterns and prohibit the usage of harmful patterns. 

I assume that when the client annotated its operations with `@atomic`, he wants a form of atomic "transactions". In that case, I think that the caller SHOULD wait for the ack. `Atomic` without `ack` does not give any value to the caller. It is tough when there are N operations in one `Atomic` update. How will the caller validate if this batch was indeed executed correctly? He would need to perform N GET queries for all mutations in the `atomic` batch and wait for the result. Sure, it is feasible but far from ideal.
When prohibiting `async` and `atomic` we are enforcing good usage patterns of our clients. 

Please note that `atomic` =! `bulk` inserts. The current implementation allows bulk inserts (N records) with `@async` directive. Wdyt?The accepted field is relevant and fulfilled with data only when used with the @%s directive.
wdyt?

Comment (ivansenic) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java:71]: I think there is bunch of use cases where I don't care about the result of the atomic operation.. So `@atomic` & `@async` makes sense for me from the developer point of view.. It's his choice at the end..

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (ivansenic) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java:166]: `atomicDirective` -> `directive` naming?

**Review by ivansenic — APPROVED** [review]
Seems good to me, let's just resolve this atomic & async discussion in proper way.. Rest is fine imo..

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java:188]: Should this comment say `accepted=true`?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/DmlSchemaBuilder.java
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/DmlSchemaBuilder.java:179]: minor: `... apply the mutations in an Async way.` --> `... apply the mutations asynchronously without waiting for the result` ?

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java]: It seems that I am out-voted here :D 
Implemented `async` AND `atomic`:
https://github.com/stargate/stargate/pull/940/commits/49746aca90892df8674f9409bb8632e8bc077d52

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (dimas-b) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java]: Yeah, I think the GraphQL infrastructure takes over HTTP code generation, so it would probably be an overkill for us to try to override the HTTP code in this PR.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java
Comment (mpenick) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/BulkMutationFetcher.java]: > I believe C* provides atomic batches in the sense that all of none of the mutations in the batch take effect eventually

**Review by mpenick — APPROVED** [review]

**Review by dimas-b — APPROVED** [review]

File: testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql
Comment (tomekl007) [reply testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql:1]: @olim7t, do you want me to wait until your PR is merged, or can I merge and you will adapt yours?

**Review by dougwettlaufer — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java:181]:
```suggestion
   * accepted=true without waiting for the result.
```

**Review by olim7t — APPROVED** [review]
👍 
I'm not convinced that this will be used very often, but the implementation looks good.

File: testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql
Comment (olim7t) [reply testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql:1]: Whoops, I just merged it 😬  So the former...

Comment (tomekl007) [reply testing/src/main/graphql/betterbotz/async/BulkInsertProductsWithAsync.graphql]: Ok, I've adapted to the new IT testing approach.

**Pull request merged**
